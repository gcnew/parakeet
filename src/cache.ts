
import { StringParser } from './string_combinators'

export {
    Accessor,

    cache, byPosAndKey
}


type Accessor = (
        data: any,
        position: number,
        parser: StringParser<any, any>,
        value?: any
    ) => any;


function cache(accessor: Accessor, leftReplace: boolean): <E, T>(p: StringParser<E, T>) => StringParser<E, T> {
    return <E, T>(p: StringParser<E, T>): StringParser<E, T> => st => {
        const pos = st.getPosition();
        const data = st.getData();
        const cache = accessor(data, pos, p);

        if (cache && (cache.kind === 'right' || !leftReplace)) {
            return cache;
        }

        const val = p(st);

        // WARNING: mutation - set the value in the state
        accessor(data, pos, p, val);

        return val;
    };
}

function mkMap() {
    const map: { [key: string]: any } = Object.create(null);

    map['del me'] = 1;
    delete map['del me'];

    return map;
}

const EMPTY_MAP = mkMap();

function cacheChain(f: (parser: StringParser<any, any>, data: any) => any):
    (
        get: (data: any) => any,
        set: (data: any, value: any) => void
    ) => Accessor
{
    return (get, set) => (data, pos, p, val) => {
        const cache = get(data) || (val ? mkMap() : EMPTY_MAP);
        const entry = cache[pos];

        // getter
        if (val === undefined) {
            if (!entry) {
                return undefined;
            }

            const key = f(p, data);
            if (Array.isArray(entry)) {
                for (const pair of entry) {
                    if (pair.key === key) {
                        return pair.value;
                    }
                }

                return undefined;
            }

            return entry.key === key ? entry.value
                                     : undefined;
        }

        // setter
        const key = f(p, data);
        const kv = { key, value: val };
        if (!entry) {
            cache[pos] = kv;
            set(data, cache);
            return;
        }

        const arr = Array.isArray(entry) ? entry : [entry];
        for (const pair of arr) {
            if (pair.key === key) {
                if (pair.value.kind === 'right') {
                    throw new Error('ASSERT: right replacement encountered');
                }

                pair.value = val;
                return;
            }
        }

        arr.push(kv);
        cache[pos] = arr;
        return;
    };
}

export const byPosAndParser = cacheChain(x => x);

function byPosAndKey(
    get: (data: any) => any,
    set: (data: any, value: any) => any
): (key: string) => Accessor {
    return key => cacheChain(_ => key)(get, set);
}
