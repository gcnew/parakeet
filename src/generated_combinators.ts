
import { Parser, ParserStream } from './parser_combinators'

export function genericCombine<M, S extends ParserStream<M>, E, A, B>(
    parsers: Parser<M, S, E, A>[],
    f: (...values: A[]) => B
): Parser<M, S, E, B> {
    if (!parsers.length) {
        throw new Error('No parsers provided');
    }
    if (!parsers.every(x => typeof x === 'function')) {
        throw new Error('Empty parser!');
    }

    return (st) => {
        let cur = st;
        const results: any[] = [];

        for (const p of parsers) {
            const res = p(cur);
            if (res.kind === 'left') {
                return res;
            }

            cur = res.value[1];
            results.push(res.value[0]);
        }

        return { kind: 'right', value: [ f(...results), cur ] };
    };
}

export function genericAlt<M, S extends ParserStream<M>, E, T>(parsers: Parser<M, S, E, T>[]): Parser<M, S, E, T> {
    if (!parsers.length) {
        throw new Error('No parsers provided');
    }
    if (!parsers.every(x => typeof x === 'function')) {
        throw new Error('Empty parser!');
    }

    return (st) => {
        let err = undefined;
        for (const p of parsers) {
            const res = p(st);
            if (res.kind !== 'left') {
                return res;
            }

            err = res;
        }

        return err!;
    };
}

export function genericChoice<M, S extends ParserStream<M>, E1, E2, T>(
    parsers: [Parser<M, S, E1, unknown>, Parser<M, S, E2, T>][]
): Parser<M, S, E1|E2, T> {
    if (!parsers.length) {
        throw new Error('No parsers provided');
    }
    if (!parsers.every(x => typeof x[0] === 'function' && typeof x[1] === 'function')) {
        throw new Error('Empty parser!');
    }

    return (st) => {
        let ret = undefined;
        for (const pair of parsers) {
            const res = pair[0](st);

            if (res.kind === 'right') {
                return pair[1](res.value[1]);
            }

            ret = res;
        }

        return ret!;
    };
}

export function alt<M, S extends ParserStream<M>, E, T1, T2>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, E, T2>
): Parser<M, S, E, T1|T2> {
    return genericAlt<M, S, any, T1|T2>([p1, p2]);
}

export function alt3<M, S extends ParserStream<M>, E, T1, T2, T3>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, E, T3>
): Parser<M, S, E, T1|T2|T3> {
    return genericAlt<M, S, any, T1|T2|T3>([p1, p2, p3]);
}

export function alt4<M, S extends ParserStream<M>, E, T1, T2, T3, T4>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, E, T4>
): Parser<M, S, E, T1|T2|T3|T4> {
    return genericAlt<M, S, any, T1|T2|T3|T4>([p1, p2, p3, p4]);
}

export function alt5<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, E, T5>
): Parser<M, S, E, T1|T2|T3|T4|T5> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5>([p1, p2, p3, p4, p5]);
}

export function alt6<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5, T6>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, unknown, T5>,
    p6: Parser<M, S, E, T6>
): Parser<M, S, E, T1|T2|T3|T4|T5|T6> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5|T6>([p1, p2, p3, p4, p5, p6]);
}

export function alt7<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5, T6, T7>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, unknown, T5>,
    p6: Parser<M, S, unknown, T6>,
    p7: Parser<M, S, E, T7>
): Parser<M, S, E, T1|T2|T3|T4|T5|T6|T7> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5|T6|T7>([p1, p2, p3, p4, p5, p6, p7]);
}

export function alt8<M, S extends ParserStream<M>, E, T1, T2, T3, T4, T5, T6, T7, T8>(
    p1: Parser<M, S, unknown, T1>,
    p2: Parser<M, S, unknown, T2>,
    p3: Parser<M, S, unknown, T3>,
    p4: Parser<M, S, unknown, T4>,
    p5: Parser<M, S, unknown, T5>,
    p6: Parser<M, S, unknown, T6>,
    p7: Parser<M, S, unknown, T7>,
    p8: Parser<M, S, E, T8>
): Parser<M, S, E, T1|T2|T3|T4|T5|T6|T7|T8> {
    return genericAlt<M, S, any, T1|T2|T3|T4|T5|T6|T7|T8>([p1, p2, p3, p4, p5, p6, p7, p8]);
}

export function combine<M, S extends ParserStream<M>, E1, E2, T1, T2, T3>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    f: (a1: T1, a2: T2) => T3
): Parser<M, S, E1|E2, T3> {
    return genericCombine<M, S, E1|E2, any, T3>([p1, p2], f);
}

export function combine3<M, S extends ParserStream<M>, E1, E2, E3, T1, T2, T3, T4>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    f: (a1: T1, a2: T2, a3: T3) => T4
): Parser<M, S, E1|E2|E3, T4> {
    return genericCombine<M, S, E1|E2|E3, any, T4>([p1, p2, p3], f);
}

export function combine4<M, S extends ParserStream<M>, E1, E2, E3, E4, T1, T2, T3, T4, T5>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4) => T5
): Parser<M, S, E1|E2|E3|E4, T5> {
    return genericCombine<M, S, E1|E2|E3|E4, any, T5>([p1, p2, p3, p4], f);
}

export function combine5<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, T1, T2, T3, T4, T5, T6>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5) => T6
): Parser<M, S, E1|E2|E3|E4|E5, T6> {
    return genericCombine<M, S, E1|E2|E3|E4|E5, any, T6>([p1, p2, p3, p4, p5], f);
}

export function combine6<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, T1, T2, T3, T4, T5, T6, T7>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    p6: Parser<M, S, E6, T6>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5, a6: T6) => T7
): Parser<M, S, E1|E2|E3|E4|E5|E6, T7> {
    return genericCombine<M, S, E1|E2|E3|E4|E5|E6, any, T7>([p1, p2, p3, p4, p5, p6], f);
}

export function combine7<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, T1, T2, T3, T4, T5, T6, T7, T8>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    p6: Parser<M, S, E6, T6>,
    p7: Parser<M, S, E7, T7>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5, a6: T6, a7: T7) => T8
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7, T8> {
    return genericCombine<M, S, E1|E2|E3|E4|E5|E6|E7, any, T8>([p1, p2, p3, p4, p5, p6, p7], f);
}

export function combine8<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, E8, T1, T2, T3, T4, T5, T6, T7, T8, T9>(
    p1: Parser<M, S, E1, T1>,
    p2: Parser<M, S, E2, T2>,
    p3: Parser<M, S, E3, T3>,
    p4: Parser<M, S, E4, T4>,
    p5: Parser<M, S, E5, T5>,
    p6: Parser<M, S, E6, T6>,
    p7: Parser<M, S, E7, T7>,
    p8: Parser<M, S, E8, T8>,
    f: (a1: T1, a2: T2, a3: T3, a4: T4, a5: T5, a6: T6, a7: T7, a8: T8) => T9
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7|E8, T9> {
    return genericCombine<M, S, E1|E2|E3|E4|E5|E6|E7|E8, any, T9>([p1, p2, p3, p4, p5, p6, p7, p8], f);
}

export function choice<M, S extends ParserStream<M>, E1, E2, E3, T1, T2>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, E3, unknown>, Parser<M, S, E2, T2>]
): Parser<M, S, E1|E2|E3, T1|T2> {
    return genericChoice<M, S, any, E1|E2, T1|T2>([p1, p2]);
}

export function choice3<M, S extends ParserStream<M>, E1, E2, E3, E4, T1, T2, T3>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, E4, unknown>, Parser<M, S, E3, T3>]
): Parser<M, S, E1|E2|E3|E4, T1|T2|T3> {
    return genericChoice<M, S, any, E1|E2|E3, T1|T2|T3>([p1, p2, p3]);
}

export function choice4<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, T1, T2, T3, T4>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, E5, unknown>, Parser<M, S, E4, T4>]
): Parser<M, S, E1|E2|E3|E4|E5, T1|T2|T3|T4> {
    return genericChoice<M, S, any, E1|E2|E3|E4, T1|T2|T3|T4>([p1, p2, p3, p4]);
}

export function choice5<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, T1, T2, T3, T4, T5>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, E6, unknown>, Parser<M, S, E5, T5>]
): Parser<M, S, E1|E2|E3|E4|E5|E6, T1|T2|T3|T4|T5> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5, T1|T2|T3|T4|T5>([p1, p2, p3, p4, p5]);
}

export function choice6<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, T1, T2, T3, T4, T5, T6>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, unknown, unknown>, Parser<M, S, E5, T5>],
    p6: [Parser<M, S, E7, unknown>, Parser<M, S, E6, T6>]
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7, T1|T2|T3|T4|T5|T6> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5|E6, T1|T2|T3|T4|T5|T6>([p1, p2, p3, p4, p5, p6]);
}

export function choice7<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, E8, T1, T2, T3, T4, T5, T6, T7>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, unknown, unknown>, Parser<M, S, E5, T5>],
    p6: [Parser<M, S, unknown, unknown>, Parser<M, S, E6, T6>],
    p7: [Parser<M, S, E8, unknown>, Parser<M, S, E7, T7>]
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7|E8, T1|T2|T3|T4|T5|T6|T7> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5|E6|E7, T1|T2|T3|T4|T5|T6|T7>([p1, p2, p3, p4, p5, p6, p7]);
}

export function choice8<M, S extends ParserStream<M>, E1, E2, E3, E4, E5, E6, E7, E8, E9, T1, T2, T3, T4, T5, T6, T7, T8>(
    p1: [Parser<M, S, unknown, unknown>, Parser<M, S, E1, T1>],
    p2: [Parser<M, S, unknown, unknown>, Parser<M, S, E2, T2>],
    p3: [Parser<M, S, unknown, unknown>, Parser<M, S, E3, T3>],
    p4: [Parser<M, S, unknown, unknown>, Parser<M, S, E4, T4>],
    p5: [Parser<M, S, unknown, unknown>, Parser<M, S, E5, T5>],
    p6: [Parser<M, S, unknown, unknown>, Parser<M, S, E6, T6>],
    p7: [Parser<M, S, unknown, unknown>, Parser<M, S, E7, T7>],
    p8: [Parser<M, S, E9, unknown>, Parser<M, S, E8, T8>]
): Parser<M, S, E1|E2|E3|E4|E5|E6|E7|E8|E9, T1|T2|T3|T4|T5|T6|T7|T8> {
    return genericChoice<M, S, any, E1|E2|E3|E4|E5|E6|E7|E8, T1|T2|T3|T4|T5|T6|T7|T8>([p1, p2, p3, p4, p5, p6, p7, p8]);
}
