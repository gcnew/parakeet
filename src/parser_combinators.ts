
export {
    /* Types */
    Parser, ParserStream, EosReached, EosExpected,

    /* combinators */
    map, mapError,

    many, oneOrMore, pwhile,

    separated, separatedZero, separatedTrailing, separatedZeroTrailing,

    peek, maybe, trai, satisfy,

    any, eos, not, terminated, pconst, pfail, forward,

    getData, setData, modifyData, withFilter,

    /* monadic binds */
    recover, inspect,

    /* Auxiliary types */
    Left, Right, Either, Literal,

    /* helpers */
    left, right, pair,
    _1, _2, _3, _4, tagged
}

import { combine } from './generated_combinators'
export * from './generated_combinators'

interface Left<L>  { kind: 'left',  value: L }
interface Right<R> { kind: 'right', value: R }

type Either<L, R> = Left<L> | Right<R>

type Literal = boolean | string | number | null | undefined | {};

interface ParserStream<S> {
    getData(): any;
    setData(data: any): this;

    next(this: this): [S, this]|null;
}

interface Parser<M, S extends ParserStream<M>, E, T> {
    (st: S): Either<E, [T, S]>
}

interface EosReached  { kind: 'pc_error', code: 'eos_reached'  }
interface EosExpected { kind: 'pc_error', code: 'eos_expected' }

function left<L>(value: L): Left<L> {
    return { kind: 'left', value };
}

function right<R>(value: R): Right<R> {
    return { kind: 'right', value };
}

function pair<F extends Literal, S>(fst: F, snd: S): [F, S] {
    return [fst, snd];
}

function _1<T>(x: T) {
    return x;
}

function _2<T>(_: any, x: T) {
    return x;
}

function _3<T>(_1: any, _2: any, x: T) {
    return x;
}

function _4<T>(_1: any, _2: any, _3: any, x: T) {
    return x;
}

function tagged<T extends string, V>(tag: T, value: V) {
    return { tag, value };
}

function mkFilterStream<S extends ParserStream<any>, T>(filter: (x: T) => T, st: S): S {
    const wrapped: S = Object.create(st);
    wrapped.next = () => {
        const val = st.next();
        if (!val) {
            return val;
        }
        return [filter(val[0]), mkFilterStream(filter, val[1])];
    };

    return wrapped;
}

function unwrapFilterStream<S extends ParserStream<any>>(st: S): S {
    const wrapped: S = Object.getPrototypeOf(st);
    if (!(wrapped.next as any)) {
        throw new Error('Assert: Unwrapping failed');
    }

    return wrapped;
}

function map<M, S extends ParserStream<M>, E, A, B>(p: Parser<M, S, E, A>, f: (a: A) => B): Parser<M, S, E, B> {
    return (st) => {
        const res = p(st);
        if (res.kind === 'left') {
            return res;
        }

        return right(pair(f(res.value[0]), res.value[1]));
    };
}

function mapError<M, S extends ParserStream<M>, E1, E2, T>(p: Parser<M, S, E1, T>, f: (e: E1) => E2): Parser<M, S, E2, T> {
    return (st) => {
        const res = p(st);
        if (res.kind === 'left') {
            return left(f(res.value));
        }

        return res;
    };
}

function withFilter<M, S extends ParserStream<M>, E, T>(
    filter: (x: M) => M,
    p: Parser<M, S, E, T>
): Parser<M, S, E, T> {
    return st => {
        const res = p(mkFilterStream(filter, st));
        if (res.kind === 'left') {
            return res;
        }

        return right(pair(res.value[0], unwrapFilterStream(res.value[1])));
    };
}

const eosReached = left<EosReached>({ kind: 'pc_error', code: 'eos_reached' });

function any<M, S extends ParserStream<M>>(st: S): Either<EosReached, [M, S]> {
    const val = st.next();

    if (!val) {
        return eosReached;
    }

    return right(val);
}

const eosExpected = left<EosExpected>({ kind: 'pc_error', code: 'eos_expected' });

function eos<S extends ParserStream<unknown>>(st: S): Either<EosExpected, [undefined, S]> {
    if (st.next()) {
        return eosExpected;
    }

    return right(pair(undefined, st));
}

function terminated<M, S extends ParserStream<M>, E, T>(p: Parser<M, S, E, T>) {
    return combine(p, eos, x => x);
}

function forward<M, S extends ParserStream<M>, E, T>(f: () => Parser<M, S, E, T>): Parser<M, S, E, T> {
    return (st) => f()(st);
}

function pconst<T extends Literal>(x: T) {
    return <S>(st: S): Either<never, [T, S]> => {
        return right(pair(x, st));
    };
}

function pfail<E extends Literal>(e: E) {
    const error = left(e);
    return <S>(_: S): Either<E, never> => {
        return error;
    };
}

function not<M, S extends ParserStream<M>, E, T>(p: Parser<M, S, E, T>): Parser<M, S, T, E> {
    return (st) => {
        const res = p(st);

        if (res.kind === 'left') {
            return right(pair(res.value, st));
        }

        return left(res.value[0]);
    };
}

function getData<S extends ParserStream<any>>(st: S): Either<never, [any, S]> {
    return right(pair(st.getData(), st));
}

function setData(data: any) {
    return <S extends { setData(data: any): any; }>(st: S): Either<never, [undefined, S]> => {
        return right(pair(undefined, st.setData(data)));
    };
}

function modifyData(f: (data: any) => any) {
    return <S extends { setData(data: any): any; getData(): any; }>(st: S): Either<never, [any, S]> => {
        const newData = f(st.getData());
        return right(pair(newData, st.setData(newData)));
    };
}

function many<M, S extends ParserStream<M>, A>(p: Parser<M, S, unknown, A>): Parser<M, S, never, A[]> {
    return (st) => {
        let cur = st;
        const results: A[] = [];

        while (true) {
            const res = p(cur);

            if (res.kind === 'left') {
                break;
            }

            cur = res.value[1];
            results.push(res.value[0])
        }

        return right(pair(results, cur));
    };
}

function pwhile<M, S extends ParserStream<M>, E, T>(
    cond: Parser<M, S, unknown, unknown>,
    body: Parser<M, S, E, T>
): Parser<M, S, E, T[]> {
    return (st) => {
        let cur = st;
        const results: T[] = [];

        while (true) {
            const temp = cond(cur);

            if (temp.kind !== 'right') {
                break;
            }

            const res = body(temp.value[1]);
            if (res.kind === 'left') {
                return res;
            }

            cur = res.value[1];
            results.push(res.value[0])
        }

        return right(pair(results, cur));
    };
}

function oneOrMore<M, S extends ParserStream<M>, E, A>(p: Parser<M, S, E, A>): Parser<M, S, E, A[]> {
    return combine(p, many(p), (x, xs) => (xs.unshift(x), xs));
}

function separated<M, S extends ParserStream<M>, E, T>(
    p: Parser<M, S, E, T>,
    sep: Parser<M, S, unknown, unknown>
): Parser<M, S, E, T[]> {
    return combine(p, pwhile(sep, p), (x, xs) => (xs.unshift(x), xs));
}

function separatedZero<M, S extends ParserStream<M>, E, T>(
    p: Parser<M, S, E, T>,
    sep: Parser<M, S, unknown, unknown>
): Parser<M, S, E, T[]> {
    return map(
        trai(p, pwhile(sep, p), (x, xs) => (xs.unshift(x), xs)),
        res => res || []
    );
}

function separatedTrailing<M, S extends ParserStream<M>, E, T>(
    p: Parser<M, S, E, T>,
    sep: Parser<M, S, unknown, unknown>
): Parser<M, S, E, T[]> {
    return combine(
        p,
        trai(
            sep,
            separatedZeroTrailing(p, sep),
            _2
        ),

        (x, xs) => xs ? (xs.unshift(x), xs) : [x]
    );
}

function separatedZeroTrailing<M, S extends ParserStream<M>, E, T>(
    p: Parser<M, S, E, T>,
    sep: Parser<M, S, unknown, unknown>
): Parser<M, S, E, T[]> {
    return combine(
        many(combine(p, sep, _1)),
        maybe(p),
        (xs, x) => (x && xs.push(x), xs)
    );
}

function peek<M, S extends ParserStream<M>, E, T>(p1: Parser<M, S, E, T>): Parser<M, S, E, T> {
    return (st) => {
        const res = p1(st);

        if (res.kind === 'left') {
            return res;
        }

        return right(pair(res.value[0], st));
    };
}

function maybe<M, S extends ParserStream<M>, A>(p1: Parser<M, S, unknown, A>): Parser<M, S, never, A|undefined> {
    return (st) => {
        const res = p1(st);

        if (res.kind === 'left') {
            return right(pair(undefined, st));
        }

        return res;
    };
}

function trai<M, S extends ParserStream<M>, E1, E2, A, B, C>(
    p1: Parser<M, S, E1, A>,
    p2: Parser<M, S, E2, B>,
    f: (a: A, b: B) => C
): Parser<M, S, E2, C|undefined> {
    return (st) => {
        const res = p1(st);

        if (res.kind === 'left') {
            return right(pair(undefined, st));
        }

        const res2 = p2(res.value[1]);
        if (res2.kind === 'left') {
            return res2;
        }

        return right(pair(f(res.value[0], res2.value[0]), res2.value[1]));
    };
}

function recover<M, S extends ParserStream<M>, E1, E2, A, B>(
    p1: Parser<M, S, E1, A>,
    f: (e: E1) => Parser<M, S, E2, B>
): Parser<M, S, E2, A|B> {
    return (st) => {
        const res = p1(st);
        if (res.kind === 'right') {
            return res;
        }

        return f(res.value)(st);
    };
}

function inspect<M, S extends ParserStream<M>, E1, E2, A, B>(
    p1: Parser<M, S, E1, A>,
    f: (e: A) => Parser<M, S, E2, B>
): Parser<M, S, E1|E2, B> {
    return (st) => {
        const res = p1(st);
        if (res.kind === 'left') {
            return res;
        }

        return f(res.value[0])(res.value[1]);
    };
}

function satisfy<M, S extends ParserStream<M>, E>(f: (x: M) => boolean, e: E): Parser<M, S, EosReached|E, M> {
    const error = left(e);
    return (st) => {
        const res = st.next();
        if (!res) {
            return eosReached;
        }

        if (!f(res[0])) {
            return error;
        }

        return right(res);
    };
}
