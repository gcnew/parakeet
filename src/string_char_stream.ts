
import { pair } from './parser_combinators'
import { TextStream } from './string_combinators'

export { StringCharStream }

type LineOffsetTable = [number, string][];

class StringCharStream implements TextStream {
    protected constructor(
        private source: string,
        private offset: number,
        private lineOffsetTable: [number, string][]
    ) {
    }

    protected newInstance(source: string, offset: number, lineOffsetTable: [number, string][]): this {
        return new StringCharStream(source, offset, lineOffsetTable) as this;
    }

    public static newInstance(source: string) {
        return new StringCharStream(source, 0, parseLineOffsets(source));
    }

    next() {
        if (this.offset >= this.source.length) {
            return null;
        }

        return pair(
            this.source[this.offset],
            this.newInstance(this.source, this.offset + 1, this.lineOffsetTable)
        );
    }

    getPosition() {
        return this.offset;
    }

    getLineCol(offset: number): [number, number] {
        const res = getLineCol(offset, this.lineOffsetTable);

        if (!res) {
            throw new Error(`Invalid offset: ${offset}`);
        }

        return res;
    }
}

function getLineCol(offset: number, lineOffsetTable: LineOffsetTable): [number, number]|null {
    let idx = binarySearch(lineOffsetTable, x => x[0] < offset ? -1 :
                                                 x[0] > offset ?  1 : 0);

    if (idx === false) {
        return null;
    }

    if (idx < 0) {
        idx = -idx - 1;
    }

    return [idx, offset - lineOffsetTable[idx][0]];
}

function parseLineOffsets(source: string): LineOffsetTable {
    const lines = source.split('\n');

    let acc = [];
    let offset = 0;
    for (const l of lines) {
        acc.push(pair(offset, l));
        offset += l.length + 1;
    }

    return acc;
}

function binarySearch<T>(arr: T[], compare: (x: T) => -1|0|1): false|number {
    let low = 0;
    let high = arr.length - 1;

    if (!arr.length) {
        return false;
    }

    while (low <= high) {
        const m = low + ((high - low) >> 1);
        const cmp = compare(arr[m]);

        if (cmp < 0) {
            low = m + 1;
            continue;
        }

        if (cmp > 0) {
            high = m - 1;
            continue;
        }

        return m;
    }

    return -low;
}
