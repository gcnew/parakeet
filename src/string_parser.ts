
import { pair } from './parser_combinators'
import { TextStream } from './string_combinators'

export { CharStream }


class CharStream implements TextStream {
    constructor(
        private source: string,
        private offset: number
    ) {
    }

    protected newInstance(source: string, offset: number): this {
        return new CharStream(source, offset) as this;
    }

    next() {
        if (this.offset >= this.source.length) {
            return null;
        }

        return pair(
            this.source[this.offset],
            this.newInstance(this.source, this.offset + 1)
        );
    }

    getPosition() {
        return this.offset;
    }

    getLineCol(offset: number) {
        // TODO: ...
        return null as any;
    }
}
