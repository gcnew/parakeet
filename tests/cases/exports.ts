
import { test } from '../test_util'

import { many, combine3, anyChar, ws, number } from '../../src'

const p = combine3(
    anyChar,
    many(ws),
    number,

    (chr, _, num) => ({ chr, num: Number(num) })
);

test(p, "s 10");
