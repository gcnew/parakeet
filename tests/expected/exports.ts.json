{
    "kind": "right",
    "value": [
        {
            "chr": "s",
            "num": 10
        },
        {
            "source": "s 10",
            "offset": 4,
            "data": {
                "lineOffsetTable": [
                    [
                        0,
                        "s 10"
                    ]
                ]
            }
        }
    ]
}
